src/watchdog/observers/fsevents.py

```diff
@@ -67,86 +71,125 @@ class FSEventsEmitter(EventEmitter):
+        self.filesystem_view = {}
 
     def on_thread_stop(self):
         if self.watch:
             _fsevents.remove_watch(self.watch)
             _fsevents.stop(self)
             self._watch = None
+        self.filesystem_view.clear()
+
+    def queue_event(self, event):
+        logger.info("queue_event %s", event)
+        EventEmitter.queue_event(self, event)
+
+    def queue_events(self, timeout, events):
+        i = 0
+        while i < len(events):
+            event = events[i]
+            logger.info(event)
+            src_path = self._encode_path(event.path)
+
+            """
+            FIXME: It is not enough to just de-duplicate the events based on
+                   whether they are coalesced or not. We must also take into
+                   account old and new state. As such we need to track all
+                   events that occurred so that we can make a correct decision
+                   about which events should be generated.
+            """
+
+            # For some reason the create and remove flags are sometimes also
+            # set for rename and modify type events, so let those take
+            # precedence.
+            if event.is_renamed:
+                # Internal moves appears to always be consecutive in the same
+                # buffer and have IDs differ by exactly one (while others
+                # don't) making it possible to pair up the two events coming
+                # from a singe move operation. (None of this is documented!)
+                # Otherwise, guess whether file was moved in or out.
+                # TODO: handle id wrapping
+                if (i + 1 < len(events) and events[i + 1].is_renamed
+                        and events[i + 1].event_id == event.event_id + 1):
+                    logger.info("Next event for rename is %s", events[i + 1])
+                    cls = DirMovedEvent if event.is_directory else FileMovedEvent
+                    dst_path = self._encode_path(events[i + 1].path)
+                    self.queue_event(cls(src_path, dst_path))
+                    self.queue_event(DirModifiedEvent(os.path.dirname(src_path)))
+                    self.queue_event(DirModifiedEvent(os.path.dirname(dst_path)))
+                    for sub_event in generate_sub_moved_events(src_path, dst_path):
+                        logger.info("Generated sub event: %s", sub_event)
+                        self.queue_event(sub_event)
+                    i += 1
+                elif os.path.exists(event.path):
+                    cls = DirCreatedEvent if event.is_directory else FileCreatedEvent
+                    self.queue_event(cls(src_path))
+                    self.queue_event(DirModifiedEvent(os.path.dirname(src_path)))
+                else:
+                    cls = DirDeletedEvent if event.is_directory else FileDeletedEvent
+                    self.queue_event(cls(src_path))
+                    self.queue_event(DirModifiedEvent(os.path.dirname(src_path)))
```

Comment:
Do we also need to generate deleted events for entire tree here? Is this tested somewhere?

Reply:
It doesn't look like it, none of the emitters is doing this explicitly. I've only found synthetic events for move and create.

Reply:
Oh, ok. The old fsevents emitter was reporting deleted events for the entire tree (a natural consequence of the snapshot diff) but if this change brings it more in line with other backends, all the better. We'll definitely need to remove all children from `filesystem_view` though, to catch future created events for those paths.

Reply:
Indeed, the `filesystem_view` is going to be problematic with those move events. Ultimately it's getting very close to be as badly performing as the old solution using directory snapshots.

That said, if we remove the filesystem view then the only weirdness that the current tests observe is the extra `FileCreatedEvent` in `test_emitter.py:test_modify`. And since this is only because the tests run so fast that they cause an the events to be coalesced then I'm inclined to simply add an expected `FileCreatedEvent` event check in the test case.

Reply:
I think having those extra events is an acceptable trade off to prevent the complexities of tracking the file system state.